File: solution/948. Bag of Tokens/948.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end());
            
        int score = 0;
        int i = 0, j = tokens.size() - 1;
        
        while (i <= j) {
            while (i <= j && power >= tokens[i]) {
                power -= tokens[i++];
                ++score;
            }
            if (score == 0 || i >= j) break;
            power += tokens[j--];
            --score;
        }
        
        return score;
    }
};

// 200 -> 100 -> 500

// 0 -> 1 -> 0 -> 1 -> 2
    
// 100, 200, 300, 400
//            i
//            j

File: solution/948. Bag of Tokens/948.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        
        score = 0
        i, j = 0, len(tokens) - 1
        
        while True:
            while i <= j and power >= tokens[i]:
                power -= tokens[i]
                i += 1
                score += 1
            if score == 0 or i >= j:
                break
            power += tokens[j]
            j -= 1
            score -= 1
            
        return score
    
